gen-ancestors-from-src.py

```diff
@@ -0,0 +1,237 @@
+        shell=isinstance(cmd, str),
+        capture_output=capture_output,
+        text=True,
+        check=check,
+        cwd=cwd,
+        encoding="utf-8",
+    )
+    if result.stdout:
+        logging.debug(f"Command stdout: {result.stdout}")
+    if result.stderr:
+        logging.debug(f"Command stderr: {result.stderr}")
+    result.check_returncode()
+    return result
+
+
+def split_archive_filename(filename):
+    """Split filename into base name and archive extension.
+
+    :param filename: Name of the file to split
+    :type filename: str
+    :returns: (filename_without_extension, extension),
+              or (filename, None) if no archive extension found
+    :rtype: tuple
+    """
+    for ext in ARCHIVE_EXTENSIONS:
+        if filename.lower().endswith(ext):
+            return (filename[: -len(ext)], ext)
+    return (filename, None)
+
+
+def parse_name_version(filename):
+    """Parse package name and version from a source filename.
+
+    Attempts to extract the package name and version from a source archive
+    filename by splitting on hyphens. Assumes the format is name-version.ext.
+
+    :param filename: Source archive filename to parse
+    :type filename: str
+    :returns: (package_name, version) where version may be None
+    :rtype: tuple
+    """
+    sname = filename
+    sver = None
+    fbn, ext = split_archive_filename(filename)
+    if ext:
+        parts = fbn.split("-")
+        sname = "-".join(parts[:-1])
+        if not sname:
+            sver = None
+            sname = fbn
+        else:
+            sver = parts[-1]
+    else:
+        logging.warning("%s is not a tarball", filename)
+    return sname, sver
+
+
+def calc_sha256_checksum(filepath, chunk_size=1024**2):
+    h = hashlib.sha256()
+    with open(filepath, "rb") as fp:
+        while True:
+            data = fp.read(chunk_size)
+            if not data:
+                break
+            h.update(data)
+    return h.hexdigest()
+
+
+def search_specfile(src_dir):
+    """Search for a specfile in the given source directory.
+
+    :param src_dir: Source directory to search in
+    :type src_dir: str
+    :returns: Path to the specfile if found
+    :rtype: str
+    :raises FileNotFoundError: If no specfile found
+    :raises OSError: If multiple specfiles found
+    """
+    specfiles = []
+    for root, _, files in os.walk(src_dir):
+        for file in files:
+            if file.endswith(".spec"):
+                specfiles.append(os.path.join(root, file))
+    if len(specfiles) == 0:
+        raise FileNotFoundError(f"No specfile found in {src_dir}")
+    elif len(specfiles) > 1:
+        raise OSError(f"Multiple specfiles found: {specfiles}")
+    return specfiles[0]
+
+
+def list_sources(specfile, srcdir="."):
+    sources = []
+    result = run_command(
+        [
+            "rpmdev-spectool",
+            "-d",
+            f"_sourcedir {srcdir}",
+            "--sources",
+            specfile,
+        ]
+    )
```

Comment:
This will likely fail on similar issues #35, should we use python3-specfile with similar hacks we have in `select_architectures.py`?